service/history/timerQueueActiveTaskExecutor.go

```diff
@@ -292,35 +299,46 @@ Loop:
 func (t *timerQueueActiveTaskExecutor) executeWorkflowTaskTimeoutTask(
 	ctx context.Context,
 	task *tasks.WorkflowTaskTimeoutTask,
-) (retError error) {
+) (isSpeculativeWorkflowTaskTimeoutTask bool, retError error) {
 	ctx, cancel := context.WithTimeout(ctx, taskTimeout)
 	defer cancel()
 
 	weContext, release, err := getWorkflowExecutionContextForTask(ctx, t.cache, task)
 	if err != nil {
-		return err
+		return false, err
 	}
 	defer func() { release(retError) }()
 
 	mutableState, err := loadMutableStateForTimerTask(ctx, weContext, task, t.metricHandler, t.logger)
 	if err != nil {
-		return err
+		return false, err
 	}
 	if mutableState == nil || !mutableState.IsWorkflowExecutionRunning() {
-		return nil
+		return false, nil
 	}
 
 	workflowTask := mutableState.GetWorkflowTaskByID(task.EventID)
 	if workflowTask == nil {
-		return nil
-	}
-	err = CheckTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), workflowTask.Version, task.Version, task)
-	if err != nil {
-		return err
+		return false, nil
 	}
 
-	if workflowTask.Attempt != task.ScheduleAttempt {
-		return nil
+	isSpeculativeWorkflowTaskTimeoutTask = workflowTask.Type == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE
```

Comment:
isSpeculativeWorkflowTaskTimeoutTask (especially for metrics) is an attribute of the task itself and should not change when the workflow task in mutable state changes. With current implementation, if there's no workflow task in the mutable state, metrics emitted will for the normal db workflow task timeout task even if the task is an in memory timer. 

If speculative workflow task can share the same eventID as normal workflow task, then when in memory timer task executes, it can timeout an existing normal workflow task before the actual timeout? Maybe speculative & normal workflow task will never share the same eventID, but I see no reason for introducing this additional complexity for reasoning the code. 

We should be able to determine if the task is an in memory timer based on the type of the executable.